funsor/distributions.py

```diff
@@ -88,68 +88,98 @@ def eager_reduce(self, op, reduced_vars):
         return super(Distribution, self).eager_reduce(op, reduced_vars)
 
     @classmethod
-    def eager_log_prob(cls, **params):
-        inputs, tensors = align_tensors(*params.values())
-        params = dict(zip(params, tensors))
+    def eager_log_prob(cls, *params):
+        inputs, tensors = align_tensors(*params)
+        params = dict(zip(cls._ast_fields, tensors))
         value = params.pop('value')
         data = cls.dist_class(**params).log_prob(value)
         return Tensor(data, inputs)
 
+    def __getattribute__(self, attr):
+        if attr in type(self)._ast_fields and attr != 'name':
+            return self.params[attr]
+        return super().__getattribute__(attr)
+
+    @classmethod
+    def _infer_value_domain(cls, **kwargs):
+        # rely on the underlying distribution's logic to infer the event_shape
+        instance = cls.dist_class(**{k: _dummy_tensor(v.output) for k, v in kwargs.items()}, validate_args=False)
```

Comment:
I think this is costly. Can we store `event_shape` information in `_wrapped_pyro_dists`?

Reply:
I'm not sure I understand your proposal, but I don't think this is unreasonably costly (note `_dummy_tensor` is always an expanded scalar), and I don't see how to implement `_infer_value_domain` generically without either using this trick or reimplementing `event_shape` inference for every underlying distribution by hand, which is what the previous version did.

Reply:
Understood! (I was confused about `event_shape` and `event_dim`).

Some distributions have additional numeric ops in their constructions. But I don't have a solution for this issue.

Reply:
I also think this is costly, triggering needless tensor ops and Python overhead and GPU-CPU communication each time a distribution is constructed. Can we enter this in a table instead? Maybe you could you move this probing logic to an on-library-load helper that populates the table of shaping metadata?

Reply:
> Can we enter this in a table instead?

The result of `_infer_value_domain` depends on the shapes and sometimes even the values (in the case of `Binomial` or `Multinomial`) of distribution parameters.  I could certainly memoize it but I'm not sure I see how to avoid constructing a new distribution at runtime for new parameter shapes - what did you have in mind?

Reply:
memoization based on shape seems fine.